convert-falcon-hf-to-gguf.py

```diff
@@ -159,22 +159,9 @@ def parse_args() -> argparse.Namespace:
 byte_decoder = {v: k for k, v in byte_encoder.items()}
 
 for i in range(vocab_size):
-    if i in reverse_vocab:
-        try:
-            text = bytearray([byte_decoder[c] for c in reverse_vocab[i]])
-        except KeyError:
-            text = bytearray()
-            for c in reverse_vocab[i]:
-                if ord(c) < 256:  # single byte character
-                    text.append(byte_decoder[ord(c)])
-                else:  # multibyte special token character
-                    text.extend(c.encode('utf-8'))
-    else:
-        print(f"Key {i} not in tokenizer vocabulary. Padding with an arbitrary token.")
-        pad_token = f"[PAD{i}]".encode("utf8")
-        text = bytearray(pad_token)
-
-    tokens.append(text)
+    tokens.append(reverse_vocab[i])
+    scores.append(0.0) # dummy
+    toktypes.append(gguf.TokenType.NORMAL)
```

Comment:
This doesn't work as-is, because scores and toktypes were removed from this file in a previous PR. Also, won't this throw KeyError now if the model's vocab_size is larger than reverse_vocab?

Reply:
And in order to satisfy the needs specified in #3405, we will need to at least provide a way (via toktypes) to differentiate between added tokens (`tokenizer.get_added_vocab()`) and normal tokens (NORMAL vs USER_DEFINED?). I would be glad to make a PR if you think that's the right way to go about this.

Reply:
> This doesn't work as-is, because scores and toktypes were removed from this file in a previous PR. Also, won't this throw KeyError now if the model's vocab_size is larger than reverse_vocab?

Until now I've only thought about the one-to-one correspondence between piece and token-id in `llama.cpp`. It would be surprising for me if this correspondence is violated in the original model. For Falcon and Aquila we seem to be good here.

Reply:
> And in order to satisfy the needs specified in #3405, we will need to at least provide a way (via toktypes) to differentiate between added tokens (`tokenizer.get_added_vocab()`) and normal tokens (NORMAL vs USER_DEFINED?). I would be glad to make a PR if you think that's the right way to go about this.

I wrote about my current understanding of added tokens [here](https://github.com/ggerganov/llama.cpp/issues/2820#issuecomment-1728361661). In short: all added tokens looked like CONTROL tokens to me for now (maybe because I have to see `sentencepiece`'s USER_DEFINED tokens in the wild yet).

Reply:
> This doesn't work as-is, because scores and toktypes were removed from this file in a previous PR.

One more remark: as far as I remember scores are needed for BPE merges. Token types NORMAL and CONTROL are used in this PR.

Reply:
The scores and toktypes were only removed because they were only placeholder values, since the GGUF spec says they are optional. If they are used for something meaningful in the future they can certainly be added back.

Should [these lines](https://github.com/goerch/llama.cpp/blob/d6d7d0f04357ff9ccb97e8b44b4eb55cf9355953/convert-falcon-hf-to-gguf.py#L150-L152) be removed, if they do not apply to Falcon? They necessarily imply that `vocab_size > len(reverse_vocab)` may be true.
```python
# The number of tokens in tokenizer.json can differ from the expected vocab size.
# This causes downstream issues with mismatched tensor sizes when running the inference
vocab_size = hparams["vocab_size"] if "vocab_size" in hparams else len(tokenizer_json["model"]["vocab"])
```

Reply:
> Should [these lines](https://github.com/goerch/llama.cpp/blob/d6d7d0f04357ff9ccb97e8b44b4eb55cf9355953/convert-falcon-hf-to-gguf.py#L150-L152) be removed, if they do not apply to Falcon? They necessarily imply that `vocab_size > len(reverse_vocab)` may be true.

I removed unused code from the conversion scripts but would not touch that code until we agree that `tokenizer.json`/`tokenizer.model` should be considered as the source of truth for the tokenizers. Furthermore I'm not sure about our strategy for the consolidation of conversion scripts (I'm thinking about moving tokenizer support to `gguf.py` in a later PR, for example).

Reply:
> This doesn't work as-is, because scores and toktypes were removed from this file in a previous PR. 

Oops, missed the impact of this completely when merging `main`: tried to repair just now.